### PR TITLE
Refactoring tpm2_createpolicy tool to support:

### PR DESCRIPTION
**Pull request by idesai**
1. Trial and Real policy sessions.
2. PCR inputs accepted as raw bytes (output of tpm2_listpcrs). Default reads current pcrs based on selection.
3. PCR selection with set-list to include multiple pcrs in the policy
4. Keep policy session alive and output policy session handle at tool exit (device tcti usage only)

### CODE REVIEW

File: lib/tpm_policy.c [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+#include <errno.h>
+#include <limits.h>
+
+#include "../lib/tpm2_util.h"
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: "tpm2_util.h" no need for path here.

File: lib/tpm_policy.c [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+#include "files.h"
+#include "log.h"
+#include "tpm_session.h"
+#include "tpm_hash.h"
+#include "tpm_policy.h"
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: alphabetical order to includes, but keep the separation you currently have between the system headers and the internal library.

File: lib/tpm_policy.c [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        case TPM_ALG_SHA384:
+            return SHA384_DIGEST_SIZE;
+            break;
+        case TPM_ALG_SHA512:
+            return SHA512_DIGEST_SIZE;
+            break;
+        case TPM_ALG_SM3_256:
+            return SM3_256_DIGEST_SIZE;
+            break;
+        default:
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: Is this case possible? Should this case actually cause an abort() because something is so wrong?

File: lib/tpm_policy.c [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    }
+}
+
+static bool evaluate_populate_pcr_digests(CREATE_POLICY_CTX *pctx, TPML_DIGEST *pcr_values) {
+    
+    uint8_t group_val=0;
+    uint8_t total_indices_for_this_alg=0;
+    uint8_t expected_pcr_input_file_size=0;
+    unsigned i, j, k, dgst_cnt=0; 
+    for (i=0; i<pctx->pcr_policy_options.pcr_selections.count; i++) {
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: space between operators.
```
(i = 0; i < pctx->pcr_policy_options.pcr_selections.count; i++)
```

File: lib/tpm_policy.c [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+static bool evaluate_populate_pcr_digests(CREATE_POLICY_CTX *pctx, TPML_DIGEST *pcr_values) {
+    
+    uint8_t group_val=0;
+    uint8_t total_indices_for_this_alg=0;
+    uint8_t expected_pcr_input_file_size=0;
+    unsigned i, j, k, dgst_cnt=0; 
+    for (i=0; i<pctx->pcr_policy_options.pcr_selections.count; i++) { 
+        for (j=0; j<pctx->pcr_policy_options.pcr_selections.pcrSelections[i].sizeofSelect; j++) {
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: same

File: lib/tpm_policy.c [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+            group_val = pctx->pcr_policy_options.pcr_selections.pcrSelections[i].pcrSelect[j];
+            total_indices_for_this_alg+=(((group_val>>3)&1)+((group_val>>2)&1)+((group_val>>1)&1)+(group_val&1));
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: Comment this, this code is not obvious. Describe what its doing and why.

File: lib/tpm_policy.c [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        }
+        expected_pcr_input_file_size += (total_indices_for_this_alg * 
+            get_size_from_alg(pctx->pcr_policy_options.pcr_selections.pcrSelections[i].hash));
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: Is there any possibility of overflow on the multiplication here?

File: lib/tpm_policy.c [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+        /*
+         * indexing with k is used as valid map while populating from input file if 
+         * file_size_test passes
+         */
+        pcr_values->count += total_indices_for_this_alg;
+        for (k=0; k<total_indices_for_this_alg; k++) {
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: spaces on operators.

File: lib/tpm_policy.c [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+            pcr_values->digests[dgst_cnt+k].t.size = 
+                get_size_from_alg(pctx->pcr_policy_options.pcr_selections.pcrSelections[i].hash);
+        }
+        dgst_cnt++;
+
+        total_indices_for_this_alg=0;
+    }
+
+    int file_size_test = (pcr_values->count == 0) ? 1:0;
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: No need for this, just use pcr_values->count == 0, since that provides you a 1 via language guarantees.

File: lib/tpm_policy.c [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    if (file_size_test && pctx->pcr_policy_options.is_raw_pcrs_file) {
+        long filesize = 0;
+        bool result = files_get_file_size(pctx->pcr_policy_options.raw_pcrs_file, &filesize);
+        if (!result) {
+                return false;
+        } else {
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: No need for an else when the if has a return... superfluous control flow.

File: lib/tpm_policy.c [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+                const char *msg = ferror(fp) ? strerror(errno) :
+                        "end of file reached";
+                LOG_ERR("Reading from file \"%s\" failed: %s",
+                        pctx->pcr_policy_options.raw_pcrs_file, msg);
+                fclose(fp);
+                return TPM_RC_NO_RESULT;
+            }
+        }
+        fclose(fp);
+    } else {
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: Big blocks of if...else statements are hard to follow. Perhaps use static internal helper functions to abstract whats going on... this is much easier to read:

```
if (a) {
  do_something_a();
} else {
  do_something_b();
}
```

Versus very long blocks of code.

File: lib/tpm_policy.c [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+        UINT32 pcr_update_counter;
+        TPML_PCR_SELECTION pcr_selection_out;
+        // Read PCRs
+        TPM_RC rval = Tss2_Sys_PCR_Read(pctx->sapi_context, 0,
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: multiple TPM_RC rval variables. Thankfully the compiler can handle this, but clean it up.

File: lib/tpm_policy.c [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+    }
+    // Calculate hashes
+    rval = tpm_hash_sequence(pctx->sapi_context, 
+        pctx->common_policy_options.policy_session->authHash, pcr_values.count, 
+        &pcr_values.digests[0], &pcr_digest);
+    if (rval != TPM_RC_SUCCESS) {
+        return rval;
+    }
+    // Call the PolicyPCR command
+    rval = Tss2_Sys_PolicyPCR(pctx->sapi_context,
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: this can be a tail call... which can be optimized by the compiler.

return Tss2_Sys_PolicyPCR();

File: lib/tpm_policy.c [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+            pctx->common_policy_options.policy_session->sessionHandle, 0, 
+            &pcr_digest, &pctx->pcr_policy_options.pcr_selections, 0);
+    //Return
+    return rval;
+}
+
+TPM_RC start_policy_session (CREATE_POLICY_CTX *pctx) {
+    if (pctx->common_policy_options.policy_session_type == TPM_SE_HMAC) {
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: Why? Comment this.

File: lib/tpm_policy.c [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+        goto build_policy_error;
+    }
+    // Get Policy Hash
+    rval = Tss2_Sys_PolicyGetDigest(pctx->sapi_context,
+            pctx->common_policy_options.policy_session->sessionHandle, 0, 
+            &pctx->common_policy_options.policy_digest, 0);
+    if (rval != TPM_RC_SUCCESS) {
+        LOG_ERR("Failed Policy Get Digest\n");
+        goto build_policy_error;
+    } else {
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: no need for else when if has goto build_policy_error

File: lib/tpm_policy.h [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+} PCR_POLICY_OPTIONS;
+
+typedef struct {
+    TSS2_SYS_CONTEXT *sapi_context;
+    COMMON_POLICY_OPTIONS common_policy_options;
+    PCR_POLICY_OPTIONS pcr_policy_options;
+} CREATE_POLICY_CTX;
+
+
+TPM_RC start_policy_session (CREATE_POLICY_CTX *pctx);
```
Comment (williamcroberts) [inline lib/tpm_policy.h]: doxygen comments on these!

File: lib/tpm_policy.c (hunk 3 above)
Comment (idesai) [inline lib/tpm_policy.c]: Good catch, returning digest size as zero and logging the error.

File: lib/tpm_policy.c (hunk 6 above)
Comment (idesai) [inline lib/tpm_policy.c]: Added inline comments.

File: lib/tpm_policy.c (hunk 7 above)
Comment (idesai) [inline lib/tpm_policy.c]: Good catch, changed data type.

File: lib/tpm_policy.c (hunk 9 above)
Comment (idesai) [inline lib/tpm_policy.c]: lingering test code, removed it.

File: lib/tpm_policy.c (hunk 14 above)
Comment (idesai) [inline lib/tpm_policy.c]: lingering test code, removed it.

File: lib/tpm_policy.h (hunk 16 above)
Comment (idesai) [inline lib/tpm_policy.h]: Added comments.

**Comment by williamcroberts** [comment]
@idesai Can you fix-up the conflicts, should be pretty trivial.

File: lib/tpm_policy.c [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+        case TPM_ALG_SHA512:
+            return SHA512_DIGEST_SIZE;
+            break;
+        case TPM_ALG_SM3_256:
+            return SM3_256_DIGEST_SIZE;
+            break;
+        default:
+            LOG_ERR("Unknown hashAlg, cannot determine digest size.\n");
+            return 0;
+            break;
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: No need for a break, when you have a return right above it.

File: lib/tpm_policy.c [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+    //Iterating the number of pcr banks selected
+    for (i=0; i < pctx->pcr_policy_options.pcr_selections.count; i++) { 
+        //Looping to check total pcr select bits in the pcr-select-octets for a bank
+        for (j=0; j < pctx->pcr_policy_options.pcr_selections.pcrSelections[i].sizeofSelect; j++) {
+            group_val = pctx->pcr_policy_options.pcr_selections.pcrSelections[i].pcrSelect[j];
+            total_indices_for_this_alg += ( ((group_val>>3) & 1) + ((group_val>>2) & 1) +
+                ((group_val>>1) &1 ) + (group_val & 1) );
+        }
+
+        dgst_size = get_size_from_alg(pctx->pcr_policy_options.pcr_selections.pcrSelections[i].hash);
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: unsigned dgst_size here, no need to declare it on line 73. Keep things scoped to smallest scope.

**Comment by idesai** [comment]
@williamcroberts per your advice, I have closed the other PRs and fragmented this PR with all the relevant commits that have dependency on the refactored tpm2_createpolicy tool.
Thanks!

File: lib/tpm_policy.h [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+//Records the type of policy and if one is selected
+typedef struct {
+        bool PolicyPCR;
+        bool is_policy_type_selected;
+}POLICY_TYPE;
+
+//Common policy options
+typedef struct {
+	SESSION *policy_session; // policy session
+	TPM_SE policy_session_type; // TPM_SE_TRIAL or TPM_SE_POLICY
```
Comment (williamcroberts) [inline lib/tpm_policy.h]: tabs here. use 4 spaces.

File: lib/tpm_policy.c [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+#include "tpm_hash.h"
+#include "tpm_session.h"
+#include "tpm_policy.h"
+#include "tpm2_util.h"
+
+static unsigned get_size_from_alg(TPMI_ALG_HASH hashAlg) {
+    switch (hashAlg) {
+        case TPM_ALG_SHA1:
+            return SHA1_DIGEST_SIZE;
+            break;
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: no break after return.

File: lib/tpm_policy.c [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+    } 
+    // And remove the session from sessions table.
+    rval = tpm_session_auth_end(pctx->common_policy_options.policy_session);
+    if (rval != TPM_RC_SUCCESS) {
+        LOG_ERR("Failed deleting session from session table\n");        
+        goto build_policy_error;
+    }
+
+build_policy_error:
+    return rval;
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: The goto doesn't do anything but return, so just replace your goto with return rval;

File: lib/tpm_policy.h [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#ifndef SRC_POLICY_H_
+#define SRC_POLICY_H_
+
+//Records the type of policy and if one is selected
+typedef struct {
+        bool PolicyPCR;
+        bool is_policy_type_selected;
```
Comment (williamcroberts) [inline lib/tpm_policy.h]: 4 spaces not 8

File: lib/tpm_policy.h [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+    TPMI_ALG_HASH policy_digest_hash_alg; // hash alg of final policy digest
+    bool extend_policy_session; // if policy session should persist
+    char policy_file[PATH_MAX]; // filepath for the policy digest
+    bool policy_file_flag; // if policy file input has been given
+    POLICY_TYPE policy_type;
+} COMMON_POLICY_OPTIONS;
+
+//pcr policy options
+typedef struct  {
+	char raw_pcrs_file[PATH_MAX]; // filepath of input raw pcrs file
```
Comment (williamcroberts) [inline lib/tpm_policy.h]: I think you borked the rebase... this should be char *raw_pcrs_file. See: https://github.com/01org/tpm2.0-tools/commit/11facc457f39659709ca29ec91327a6bb289f616

File: lib/tpm_policy.h [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+    POLICY_TYPE policy_type;
+} COMMON_POLICY_OPTIONS;
+
+//pcr policy options
+typedef struct  {
+	char raw_pcrs_file[PATH_MAX]; // filepath of input raw pcrs file  
+    TPML_PCR_SELECTION pcr_selections; // records user pcr selection per setlist 
+    bool is_set_list; // if user has provided the setlist choice  
+    bool is_raw_pcrs_file; // if user has provided a raw pcrs file for policy calc
+} PCR_POLICY_OPTIONS;
```
Comment (williamcroberts) [inline lib/tpm_policy.h]: Why are the structs caps (here and throughout), make them lowercase.

File: lib/tpm_policy.h [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+
+//policy options
+typedef struct {
+    TSS2_SYS_CONTEXT *sapi_context; // system API context from main
```
Comment (williamcroberts) [inline lib/tpm_policy.h]: This should likely be a separate parameter to the routines, since it doesn't actually relate to a policy.

File: lib/tpm_policy.h [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+    COMMON_POLICY_OPTIONS common_policy_options; 
+    PCR_POLICY_OPTIONS pcr_policy_options;
+} CREATE_POLICY_CTX;
```
Comment (williamcroberts) [inline lib/tpm_policy.h]: Maybe just call it a tpm2_policy structure? It's not really a context so to speak.

File: lib/tpm_policy.h [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * Calls start policy session, initiates specific policy type functions,
+ * saves policy digest when in TRIAL mode and flushes context at end of operation
+ * or keeps it alive depending upon the user choice.
+ * @param pctx.common_policy_options.policy_session_type
+ *  To start a TPM_SE_TRIAL or TPM_SE_POLICY policy session.
+ * @return
+ *  TPM_RC_SUCCESS on success, not otherwise.
+ */
+TPM_RC build_policy(CREATE_POLICY_CTX *pctx, TPM_RC (*build_policy_function)(CREATE_POLICY_CTX* pctx));
```
Comment (williamcroberts) [inline lib/tpm_policy.h]: the file is called tpm_policy.h, thus these routines should be called tpm2_policy_build(), tpm_policy_build_pcr(), etc. Also, im trying to get everything in lib as tpm2_<name>.h and tpm2_name.c so it's a bit more standardized.

File: tools/tpm2_createpolicy.c [hunk 28]
```diff
@@ -37,266 +37,101 @@
...
-TPM_RC build_pcr_policy(createpolicypcr_ctx *ctx) {
-    TPML_PCR_SELECTION pcrs = {
-        .count = 1, 
-        .pcrSelections[0] = {
-            .hash = ctx->hash_alg,
-            .sizeofSelect = ctx->max_supported_pcrs/8,
-        },
-    };
-    pcrs.pcrSelections[0].pcrSelect[ (ctx->pcr_index/8) ] |= ( 1 << ( (ctx->pcr_index) % 8) );
+TPM_RC parse_policy_type_specific_command (CREATE_POLICY_CTX *pctx) {
```
Comment (williamcroberts) [inline tools/tpm2_createpolicy.c]: static, looks internal only.

File: tools/tpm2_createpolicy.c [hunk 29]
```diff
@@ -37,266 +37,101 @@
...
 
     int opt;
-    bool result;
-    while ((opt = getopt_long(argc, argv, "Pf:i:g:F:", sOpts, NULL)) != -1) {
+    while ((opt = getopt_long(argc, argv, "f:g:L:F:Pae", sOpts, NULL)) != -1) {
         switch (opt) {
         case 'f':
-            ctx->policyfile = optarg;
+            pctx->common_policy_options.policy_file_flag = true;
+            snprintf(pctx->common_policy_options.policy_file,
```
Comment (williamcroberts) [inline tools/tpm2_createpolicy.c]: char * here, no more of this snprintf() copying. See: https://github.com/01org/tpm2.0-tools/commit/11facc457f39659709ca29ec91327a6bb289f616

This would re-open bug: #260

Comment (williamcroberts) [inline tools/tpm2_createpolicy.c]: Applies here any anywhere else.

File: tools/tpm2_createpolicy.c [hunk 30]
```diff
@@ -37,266 +37,101 @@
...
+                showArgError(optarg, argv[0]);
                 return false;
             }
+            pctx->pcr_policy_options.is_set_list = true;
             break;
         case 'F':
-            ctx->raw_pcr_file = optarg;
+            pctx->pcr_policy_options.is_raw_pcrs_file = true;
+            snprintf(pctx->pcr_policy_options.raw_pcrs_file, 
+                sizeof(pctx->pcr_policy_options.raw_pcrs_file), "%s", optarg);
```
Comment (williamcroberts) [inline tools/tpm2_createpolicy.c]: This goes away, just char * and assign it.
```
pctx->pcr_policy_options.raw_pcrs_file = optarg;
```

File: tools/tpm2_createpolicy.c [hunk 31]
```diff
@@ -320,27 +160,23 @@ int execute_tool(int argc, char *argv[], char *envp[], common_opts_t *opts,
...
-        .policy_digest = TPM2B_TYPE_INIT(TPM2B_DIGEST, buffer),
-        .pcr_index = 0,
-        .max_supported_pcrs = 0,
-        .hash_alg = TPM_ALG_SHA1,
-        .policyfile = NULL,
-        .raw_pcr_file = NULL,
-        .pcr_flags.policy_type_pcr_flag = false,
-        .pcr_flags.pcr_index_flag = false,
-        .pcr_flags.raw_pcr_flag = false
+    CREATE_POLICY_CTX pctx = {
```
Comment (williamcroberts) [inline tools/tpm2_createpolicy.c]: tool context structures should be private to the tool. You're exposing a an internal policy building API. Which I am assuming more than just tpm2_createpolicy() is going to use.

You need to seperate what should be a part of the polict structure, and the rest of the arguments. I don't see sapi_context applying to create_policy structure.

I think a better name would be to call it a tpm2_policy structure, and pass sapi_context as an argument to the tpm2_policy_build() routines.

File: tools/tpm2_createpolicy.c [hunk 32]
```diff
@@ -320,27 +160,23 @@ int execute_tool(int argc, char *argv[], char *envp[], common_opts_t *opts,
...
+        .sapi_context = sapi_context,
+        .common_policy_options = {
+            .policy_session = NULL,
+            .policy_session_type = TPM_SE_TRIAL,
+            .policy_digest = TPM2B_TYPE_INIT(TPM2B_DIGEST, buffer),
+            .policy_digest_hash_alg = TPM_ALG_SHA256,
```
Comment (williamcroberts) [inline tools/tpm2_createpolicy.c]: You're requiring the caller to magically know the right defaults for this. I would put this into a macro in tpm2_policy.h
```
...
.common_policy_options = TPM2_COMMON_POLICY_INIT,
},
```

File: lib/tpm_policy.c [hunk 33]
```diff
@@ -0,0 +1,269 @@
...
+        .count = 0
+    };
+
+    bool result = evaluate_populate_pcr_digests(pctx, &pcr_values);
+    if (!result) {
+        return TPM_RC_NO_RESULT;
+    } 
+
+    //If PCR input for policy is from raw pcrs file
+    if (pctx->pcr_policy_options.is_raw_pcrs_file) {
```
Comment (williamcroberts) [inline lib/tpm_policy.c]: What is the use case when you would build a policy without a policy file?

File: lib/tpm_policy.h [hunk 34]
```diff
@@ -0,0 +1,98 @@
...
+TPM_RC start_policy_session (CREATE_POLICY_CTX *pctx);
+
+/**
+ * Extends PolicyPCR specifics into the policy digest of active session
+ * Takes the input from raw pcr file or current pcr values.
+ * @param pctx.common_policy_options.policy_session_type
+ *  To start a TPM_SE_TRIAL or TPM_SE_POLICY policy session
+ * @param pctx.pcr_policy_options.is_set_list
+ *  To ensure user has selected the pcr indices across banks
+ * @param pctx.pcr_policy_options.is_raw_pcrs and .raw_pcrs_file
```
Comment (williamcroberts) [inline lib/tpm_policy.h]: @param is for argument parameters. You're describing the structure, the structure should be annotated with this information.

File: lib/tpm_policy.h [hunk 35]
```diff
@@ -0,0 +1,98 @@
...
+ *  Records if a raw pcrs file has been given as as input and captures file path
+ * @return
+ *  TPM_RC_SUCCESS on success, not otherwise.
+ */
+TPM_RC build_pcr_policy(CREATE_POLICY_CTX *pctx);
```
Comment (williamcroberts) [inline lib/tpm_policy.h]:
```
TPM_RC tpm2_policy_build_pcr(sapi_context, raw_pcr_file, tpm2_policy *policy);
```
Here and throughout. Then @param should include what sapi_context is, raw_pcr_file and policy paramters.

**Comment by williamcroberts** [comment]
Any tests for this?

File: lib/tpm_policy.h (hunk 35 above)
Comment (williamcroberts) [inline lib/tpm_policy.h]: I don't see this used externally in this PR, does this need to be public API? Perhaps make it static and pull it out when you need it?

File: lib/tpm_policy.h (hunk 27 above)
Comment (williamcroberts) [inline lib/tpm_policy.h]: I don't see this used elsewhere? Should this be public API?

File: lib/tpm_policy.h (hunk 22 above)
Comment (idesai) [inline lib/tpm_policy.h]: 👍

File: lib/tpm_policy.h (hunk 19 above)
Comment (idesai) [inline lib/tpm_policy.h]: 👍

File: lib/tpm_policy.h (hunk 23 above)
Comment (idesai) [inline lib/tpm_policy.h]: 👍

File: lib/tpm_policy.h (hunk 24 above)
Comment (idesai) [inline lib/tpm_policy.h]: 👍

File: lib/tpm_policy.h (hunk 25 above)
Comment (idesai) [inline lib/tpm_policy.h]: Separated the sapi context parameter to appear in the context structure.

File: lib/tpm_policy.h (hunk 26 above)
Comment (idesai) [inline lib/tpm_policy.h]: 👍

File: lib/tpm_policy.h (hunk 34 above)
Comment (idesai) [inline lib/tpm_policy.h]: Irrelevant since file has been removed since.

File: lib/tpm_policy.h (hunk 35 above)
Comment (idesai) [inline lib/tpm_policy.h]: Irrelevant since file has been removed since.

File: lib/tpm_policy.h (hunk 27 above)
Comment (idesai) [inline lib/tpm_policy.h]: @williamcroberts you are right. file has been removed since.

File: lib/tpm_policy.c (hunk 20 above)
Comment (idesai) [inline lib/tpm_policy.c]: 👍

File: lib/tpm_policy.c (hunk 33 above)
Comment (idesai) [inline lib/tpm_policy.c]: It is building without pcr file in this case, as in read from tpm pcrs instead.

File: lib/tpm_policy.c (hunk 21 above)
Comment (idesai) [inline lib/tpm_policy.c]: 👍

File: tools/tpm2_createpolicy.c (hunk 28 above)
Comment (idesai) [inline tools/tpm2_createpolicy.c]: 👍

File: tools/tpm2_createpolicy.c (hunk 29 above)
Comment (idesai) [inline tools/tpm2_createpolicy.c]: 👍

File: tools/tpm2_createpolicy.c (hunk 30 above)
Comment (idesai) [inline tools/tpm2_createpolicy.c]: 👍

File: tools/tpm2_createpolicy.c (hunk 31 above)
Comment (idesai) [inline tools/tpm2_createpolicy.c]: Please take a look at the fix.

File: tools/tpm2_createpolicy.c (hunk 32 above)
Comment (idesai) [inline tools/tpm2_createpolicy.c]: 👍

**Comment by idesai** [comment]
@williamcroberts not sure why the travis build is failing, the same tests pass on my vm

**Comment by williamcroberts** [comment]
Parse PCR policy API was changed to bool:
https://github.com/01org/tpm2.0-tools/commit/8e05aa3991725f6c21ce1037f5d8d61444905028

So when the build system was pulling your change ontop of master, it was failing since true was not the expected return.

```
diff --git a/tools/tpm2_createpolicy.c b/tools/tpm2_createpolicy.c
index 857f19e..0e5f362 100644
--- a/tools/tpm2_createpolicy.c
+++ b/tools/tpm2_createpolicy.c
@@ -374,7 +374,7 @@ static bool init(int argc, char *argv[], create_policy_ctx *pctx) {
             }
             break;
         case 'L':
-            if (pcr_parse_selections(optarg,
+            if (!pcr_parse_selections(optarg,
                 &pctx->pcr_policy_options.pcr_selections) != 0) {
                 showArgError(optarg, argv[0]);
                 return false;
```

**Comment by idesai** [comment]
@williamcroberts thanks! I did in fact rebase on an older commit. Good catch!
This could be pulled now.

File: Makefile.am [hunk 36]
```diff
@@ -93,7 +93,7 @@ lib_libcommon_a_SOURCES = \
     lib/tpm_hmac.c \
     lib/tpm_kdfa.c \
     lib/tpm_session.c \
-    lib/tpm2_util.c
+    lib/tpm2_util.c
```
Comment (williamcroberts) [inline Makefile.am]: Trailing whitespace added...
```
git diff --check
```
Would have caught this.

Comment (williamcroberts) [inline Makefile.am]: On your branch idsea/refactor_tpm2_createpolicy:

```
$ git diff --check origin/master 
Makefile.am:96: trailing whitespace.
+    lib/tpm2_util.c 
man/tpm2_createpolicy.8.in:39: trailing whitespace.
+File to save the policy digest. 
man/tpm2_createpolicy.8.in:57: trailing whitespace.
+The list of pcr banks and selected PCRs' ids   (0~23) for each bank 
man/tpm2_evictcontrol.8.in:63: trailing whitespace.
+Optional Input session handle from a policy session for authorization. 
tools/tpm2_dictionarylockout.c:130: trailing whitespace.
+        { "input-session-handle",required_argument,NULL,'S'}, 
```

Comment (williamcroberts) [inline Makefile.am]: Ill fix them up post merge.
